codegen/service/testdata/views_code.go

```diff
@@ -0,0 +1,303 @@
+// projected type and a view.
+func NewResultType(p *ResultTypeView, view string) *ResultType {
+	return &ResultType{
+		Projected: p,
+		View:      view,
+	}
+}
+
+// Validate runs the validations defined on ResultType.
+func (result *ResultType) Validate() (err error) {
+	projected := result.Projected
+	switch result.View {
+	case "tiny":
+		if projected.A == nil {
+			err = goa.MergeErrors(err, goa.MissingFieldError("a", "projected"))
+		}
+	default:
+		if projected.A == nil {
+			err = goa.MergeErrors(err, goa.MissingFieldError("a", "projected"))
+		}
+		if projected.B == nil {
+			err = goa.MergeErrors(err, goa.MissingFieldError("b", "projected"))
+		}
+	}
+	return
+}
+`
+
+var ResultWithUserTypeCode = `// ResultTypeView is a type used by ResultType type to project based on a view.
+type ResultTypeView struct {
+	A *UserType
+	B *string
+}
+
+// ResultType is the viewed result type that is projected based on a view.
+type ResultType struct {
+	// Type to project
+	Projected *ResultTypeView
+	// View to render
+	View string
+}
+
+// UserType is a type that runs validations on a projected type.
+type UserType struct {
+	A *string
+}
+
+// NewResultType initializes ResultType viewed result type from ResultTypeView
+// projected type and a view.
+func NewResultType(p *ResultTypeView, view string) *ResultType {
+	return &ResultType{
+		Projected: p,
+		View:      view,
+	}
+}
+
+// Validate runs the validations defined on ResultType.
+func (result *ResultType) Validate() (err error) {
+	projected := result.Projected
+	switch result.View {
+	case "tiny":
+		if projected.A == nil {
+			err = goa.MergeErrors(err, goa.MissingFieldError("a", "projected"))
+		}
+	default:
+		if projected.A == nil {
+			err = goa.MergeErrors(err, goa.MissingFieldError("a", "projected"))
+		}
+	}
+	return
+}
+`
+
+const ResultWithResultTypeCode = `// RTView is a type used by RT type to project based on a view.
+type RTView struct {
+	A *string
+	B *RT2
+	C *RT3
+}
+
+// RT is the viewed result type that is projected based on a view.
+type RT struct {
+	// Type to project
+	Projected *RTView
+	// View to render
+	View string
+}
+
+// RT2View is a type used by RT2 type to project based on a view.
+type RT2View struct {
+	C *string
+	D *UserType
+	E *string
+}
+
+// RT2 is the viewed result type that is projected based on a view.
+type RT2 struct {
+	// Type to project
+	Projected *RT2View
+	// View to render
+	View string
+}
+
+// UserType is a type that runs validations on a projected type.
+type UserType struct {
+	P *string
+}
+
+// RT3View is a type used by RT3 type to project based on a view.
+type RT3View struct {
+	X []string
+	Y map[int]*UserType
+	Z *string
+}
+
+// RT3 is the viewed result type that is projected based on a view.
+type RT3 struct {
+	// Type to project
+	Projected *RT3View
+	// View to render
+	View string
+}
+
+// NewRT initializes RT viewed result type from RTView projected type and a
+// view.
+func NewRT(p *RTView, view string) *RT {
```

Comment:
I am not sure these are really needed? couldn't the decode code simply create the struct inline?